Speculation rules tag (#39057)
mdn/content: files/en-us/web/http/reference/headers/sec-speculation-tags/index.md

<table class="properties">
  <tbody>
    <tr>
      <th scope="row">Header type</th>
      <td>{{Glossary("Request header")}}</td>
    </tr>
    <tr>
      <th scope="row">{{Glossary("Forbidden request header")}}</th>
      <td>Yes (<code>Sec-</code> prefix)</td>
    </tr>
  </tbody>
</table>

## Syntax

```http
Speculation-Rules: <tag-list>
```

## Directives

- `<tag-list>`
  - : A comma-separated list of tags indicating [Speculation Rules API](/en-US/docs/Web/API/Speculation_Rules_API) rules that may have initiated this request. See [Speculation rules JSON representation](/en-US/docs/Web/HTML/Element/script/type/speculationrules#speculation_rules_json_representation) for the syntax reference.

## Examples

### Speculation from a rule with no explicit tag

```html
<script type="speculationrules">
  {
    "prefetch": [
      {
        "urls": ["next.html", "next2.html"]
      }
    ]
  }
</script>
```

If a speculation happens from speculation rule with no tag then `null` is sent in the `Speculation-Rules` header.

```http
Speculation-Rules: null
```

### Speculation from a rule with a tag

```html
<script type="speculationrules">
  {
    "prefetch": [
      {

>         "tag": "my-rules",

Reviewer: ```suggestion
        "tag": "my-rule",
```

        "urls": ["next.html", "next2.html"]
      }
    ]
  }
</script>
```

If a speculation happens from speculation rule with a tag then the tag name is sent in the `Speculation-Rules` header.

```http

> Speculation-Rules: "my-rules"

Author: ```suggestion
Speculation-Rules: "my-rule"
```

```

### Speculation from multiple rules

```html
<script type="speculationrules">
  {
    "prefetch": [
      {

>         "tag": "my-rules",

Reviewer: ```suggestion
        "tag": "my-rule",
```

        "urls": ["next.html", "next2.html"],
        "eagerness": "moderate"
      }
    ]
  }
</script>
<script type="speculationrules">
  {
    "prefetch": [
      {

>         "tag": "cdn-rules",

Reviewer: ```suggestion
        "tag": "cdn-rule",
```

        "urls": ["next.html", "next.html"],
        "eagerness": "conservative"
      }
    ]
  }
</script>
```

> If a speculation is triggered by multiple speculation rules then all tags will be included. For example if the speculation is initiated with the by hovering over the link (`"eagerness": "moderate"`), then only the `my-rules` tag will be sent:

Author: ```suggestion
In this example, if the speculation is initiated by the user hovering over the link for 200 milliseconds (`"eagerness": "moderate"`), then only the `my-rule` tag will be sent in the header:
```

```http

> Speculation-Rules: "my-rules"

Author: ```suggestion
Speculation-Rules: "my-rule"
```

```

But if the link is clicked immeadiately, without waiting for the 200 milliseconds hover, then both rules would have triggered a speculation so both tags will be includes in the header:

```http

> Speculation-Rules: "my-rules" "cdn-rules"

Author: ```suggestion
Speculation-Rules: "my-rule" "cdn-rule"
```

```

### Speculation from multiple rules wiht and without tags

```html
<script type="speculationrules">
  {
    "prefetch": [
      {
        "urls": ["next.html", "next2.html"],
        "eagerness": "moderate"
      }
    ]
  }
</script>
<script type="speculationrules">
  {
    "prefetch": [
      {
        "tag": "cdn-rules",
        "urls": ["next.html", "next.html"],
        "eagerness": "conservative"
      }
    ]
  }
</script>
```

Similar to the previous example, if the link is clicked immeadiately, without waiting for the 200 milliseconds hover, then both rules would have triggered a speculation so both tags will be includes in the header and `null` will be sent for any rules without a tag:

```http

> Speculation-Rules: null "cdn-rules"

Author: ```suggestion
Speculation-Rules: null "cdn-rule"
```

> ```

Author: Definitely worth explaining when both tags are used. Added an example in https://github.com/mdn/content/pull/39057/commits/d318a3ed8e8f91bf6e49a62d3ef6439afc040922
